### ISSUE TITLE
bean-report hledger/ledger puts link/tag comments before payee/description

### ISSUE THREAD

**Issue by blais** [beancount/beancount#199]
**[Original report](https://bitbucket.org/blais/beancount/issue/199) by Trevor Davis (Bitbucket: [trevorld](https://bitbucket.org/trevorld), GitHub: [trevorld](https://github.com/trevorld)).**

----------------------------------------

Thanks for your awesome beancount package!

I've come across a bug in the bean-report hledger/ledger export for transactions that have tags or links.  In such cases the link/tag comments are placed before the payee/description field in the resulting ledger/hledger files.  For example

```
2017-01-01 * "Payee" "Description" ^link #tag
    Assets:Cash             200 USD
    Equity:Opening
```

gets exported to (H)Ledger as

```
2017-01-01 * ;; Link: ^link ;; Tag: #tag Payee | Description
    Assets:Cash             200 USD
    Equity:Opening
```

instead of 

```
2017-01-01 *  Payee | Description ;; Link: ^link ;; Tag: #tag
    Assets:Cash             200 USD
    Equity:Opening
```

In particular when I import my beancount file into an R data frame via hledger csv export for further querying this bug causes my payee / description info to be lost for any transactions with a tag or link.  If I instead import my beancount file into R via ledger csv export the info isn't fully lost but it is much much harder to further parse into payee, description, and tag/link metadata fields than if the bug was fixed.  I currently don't want to import my beancount file into R via  `bean-query -f csv filename.beancount "SELECT *"` since it seems to be missing the critical account field (as well as tag/link metadata fields).  

For your convenience I've attached a minimal beancount file that includes four transactions that should highlight this bug.

```
bean-report min_reproducible_example.beancount hledger > mre.hledger
bean-report min_reproducible_example.beancount ledger > mre.ledger
hledger -f mre.hledger register -Ocsv
ledger -f mre.ledger csv
bean-query -f csv min_reproducible_example.beancount "SELECT *"
```

Thanks for your attention to this matter!

**Comment by blais** [comment]
**Original comment by Trevor Davis (Bitbucket: [trevorld](https://bitbucket.org/trevorld), GitHub: [trevorld](https://github.com/trevorld)).**

----------------------------------------

Link to the bug in my downstream R package project:

https://github.com/trevorld/ledgeR/issues/1

**Comment by blais** [comment]
**Original comment by Trevor Davis (Bitbucket: [trevorld](https://bitbucket.org/trevorld), GitHub: [trevorld](https://github.com/trevorld)).**

----------------------------------------

Although getting the tag/link metadata formatting correct is less critical then getting the Payee/Description out of the transaction comments I later noticed another minor bug in `bean-report hledger`: multiple metadata 'tags' on one line should ideally be separated by a comma `,` and not a double semi-colon `;;`  i.e. for hledger: 


```
2016/01/10 * ;; Tag: #tag ;; Link: ^link Restaurant | Restaurant
  Expenses:Food:Restaurant                                                20.07 USD
  Liabilities:JT-Credit-Card                                             -20.07 USD

```

should ideally be something like


```
2016/01/10 * Restaurant | Restaurant ;; Tag: #tag , Link: ^link 
  Expenses:Food:Restaurant                                                20.07 USD
  Liabilities:JT-Credit-Card                                                    -20.07 USD

```

Otherwise hledger doesn't treat each metadata 'tag' as a separate tag but instead treats the following metadata as part of whatever tag came first and in cases where someone used both a #tag and a ^link in the original beancount file it could fail to filter on whichever metadata came second in the transaction comment i.e. in the first transaction in the minimal example the ^link1 came after the #tag1 so hledger fails to filter on it correctly because the two metadata were separated by a `;;` instead of a `,` and so thinks there is only one metadata tag `Tag` with the value `#tag1 ;; ^link1` instead of two metadata tags `Tag` with the value `#tag1` and `Link` with the value `^link1`
```
$ bean-report min_reproducible_example.beancount hledger | hledger -f-  register -Ocsv tag:Tag=tag1
"txnidx","date","description","account","amount","total"
"1","2017/01/01","","Assets:Cash","200 USD","200 USD"
"1","2017/01/01","","Equity:Opening","-200 USD","0"
$ bean-report min_reproducible_example.beancount hledger | hledger -f-  register -Ocsv tag:Link=link1
"txnidx","date","description","account","amount","total"

```

http://hledger.org/manual.html#comments

**Comment by blais** [comment]
**Original comment by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

for the record, tags appear to be different in ledger and hledger.

```
13:56 < tbm> sm: so it seems tags are different in hledger.  ledger: :tag: vs heldger tag:
13:57 < tbm> sm: oh
14:00 < tbm> sm: yeah, the syntax is different
14:12 < sm> tbm: yes. I think ledger later added tag: as well as :tag:
14:14 < tbm> sm: but they are slightly incompatible :/
14:15 < sm> sure, as in various other things.. is it causing pain ?
14:17 < tbm> sm: incompatibilities can always cause pain imho.  I just noticed it because I want to fix 
             https://bitbucket.org/blais/beancount/issues/199/bean-report-hledger-ledger-puts-link-tag
14:21 < sm> bean-report is trying to generate one syntax that works the same in bth ledger and hledger I guess ?
14:22 < tbm> sm: beanrn-report has both a 'ledger' and a hledger' command so I think they could be different
14:22 < tbm> sm: is there a preference for tag1:, tag2: for hledger over :tag1:tag2:?
14:25 < sm> tbm: that's good then, it sounds like just a matter of generating correct hledger syntax
14:25 < sm> hledger doesn't support :tag:, only tag:
14:26 < tbm> sm: really? :tag: worked for me in a test
14:27 < sm> well.. it'll probably ignore the leading :
14:27 < sm> if so I take it back :) but tag1:, tag2: would be preferred
14:27 < tbm> yeah, maybe
```

**Comment by blais** [comment]
**Original comment by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

I created pull request #98

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Improve output of tags and links for ledger and hleger

Tags and links were printed before the payee information
and not after.  Also, while ledger and hledger don't have
a concept of links, there is a concept of tags, so use the
right format (:tag1:tag2: for ledger, tag1:, tag2: for
hleger).

Fixes #199
